File: server/controllers/jobBoardEntriesController.ts
import { jobBoardEntriesService } from "@/services/jobBoardEntriesService.js";
import type { Request, Response } from "express";

export const getJobBoardEntries = async (req: Request, res: Response) => {
  const userId = req.user!.id;
  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  const jobBoardEntries = await jobBoardEntriesService.getJobBoardEntries(userId);
  return res.status(200).json({ jobBoardEntries });
};

export const getJobBoardEntry = async (req: Request, res: Response) => {
  const userId = req.user!.id;
  const { id } = req.params;

  const idNumber = parseInt(id as string);
  const jobBoardEntry = await jobBoardEntriesService.getJobBoardEntry(userId, idNumber);
  return res.status(200).json({ jobBoardEntry });
};

export const createJobBoardEntry = async (req: Request, res: Response) => {
  const { title, company, location, salary, url, description, status } = req.body;
  const userId = req.user!.id;

  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  const jobBoardEntry = await jobBoardEntriesService.createJobBoardEntry(userId, title, company, location, salary, url, description, status);
  return res.status(200).json({ jobBoardEntry });
};

export const bulkImportJobBoardEntries = async (req: Request, res: Response) => {
  const { entries } = req.body;
  const userId = req.user!.id;

  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  if (!Array.isArray(entries) || entries.length === 0) {
    return res.status(400).json({ error: "Entries array is required and must not be empty" });
  }

  const jobBoardEntries = await jobBoardEntriesService.bulkImportJobBoardEntries(userId, entries);
  return res.status(200).json({ jobBoardEntries });
};

export const updateJobBoardEntry = async (req: Request, res: Response) => {
  const { id, title, company, location, salary, url, description, status, number, tagNames } = req.body;
  const userId = req.user!.id;
  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  const jobBoardEntry = await jobBoardEntriesService.updateJobBoardEntry(userId, id, title, company, location, salary, url, description, status, number, tagNames);
  return res.status(200).json({ jobBoardEntry });
};

export const deleteJobBoardEntry = async (req: Request, res: Response) => {
  const { id } = req.params;
  const userId = req.user!.id;
  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  const idNumber = parseInt(id as string);
  const jobBoardEntry = await jobBoardEntriesService.deleteJobBoardEntry(userId, idNumber);
  return res.status(200).json({ jobBoardEntry });
};

export const uploadCv = async (req: Request, res: Response) => {
  const { id } = req.params;
  const { cvText } = req.body;
  const cvFile = req.file;
  const userId = req.user!.id;

  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  const idNumber = parseInt(id as string);
  const jobBoardEntry = await jobBoardEntriesService.uploadCv(
    userId,
    idNumber,
    cvText,
    cvFile
  );

  return res.status(200).json({ jobBoardEntry });
};

export const uploadCoverLetter = async (req: Request, res: Response) => {
  const { id } = req.params;
  const { coverLetterText } = req.body;
  const coverLetterFile = req.file;
  const userId = req.user!.id;

  if (!userId) {
    return res.status(400).json({ error: "User ID is required" });
  }

  const idNumber = parseInt(id as string);
  const jobBoardEntry = await jobBoardEntriesService.uploadCoverLetter(
    userId,
    idNumber,
    coverLetterText,
    coverLetterFile
  );

  return res.status(200).json({ jobBoardEntry });
};

File: backend/app/controllers/job_board_entries_controller.rb
class JobBoardEntriesController < ApplicationController
  before_action :require_authentication
  before_action :set_job_board_entry, only: %i[show update destroy upload_cv upload_cover_letter]

  UNAUTHORIZED_MESSAGES = {
    show: "You are not authorized to view this job board entry",
    update: "You are not authorized to update this job board entry",
    destroy: "You are not authorized to delete this job board entry",
    upload_cv: "You are not authorized to upload a CV for this job board entry",
    upload_cover_letter: "You are not authorized to upload a cover letter for this job board entry"
  }.freeze

  def index
    entries = current_user.job_board_entries.includes(:job_board_tags)
    render json: {
      job_board_entries: serialize(entries, with: JobBoardEntryBlueprint),
      is_empty: entries.empty?
    }
  end

  def show
    render json: { job_board_entry: serialize(@job_board_entry, with: JobBoardEntryBlueprint) }
  end

  def create
    entry = JobBoardEntry.create_for_user!(current_user, entry_attributes, tag_names: params[:tags])
    render json: { job_board_entry: serialize(entry, with: JobBoardEntryBlueprint) }
  end

  def bulk_import
    entries = params[:entries]
    unless entries.is_a?(Array) && entries.any?
      raise ApiError.new("Entries array is required and must not be empty", status: :bad_request)
    end

    created = JobBoardEntry.bulk_import_for_user!(current_user, import_attributes(entries))
    render json: { job_board_entries: serialize(created, with: JobBoardEntryBlueprint) }
  end

  def update
    @job_board_entry.update_for_user!(entry_attributes, tag_names: params[:tag_names])
    render json: { job_board_entry: serialize(@job_board_entry, with: JobBoardEntryBlueprint) }
  end

  def destroy
    snapshot = serialize(@job_board_entry, with: JobBoardEntryBlueprint)
    @job_board_entry.destroy_and_renumber!
    render json: { job_board_entry: snapshot }
  end

  def stats
    render json: serialize(JobBoardEntry.stats_for(current_user), with: JobBoardEntriesStatsBlueprint)
  end

  def upload_cv
    @job_board_entry.upload_cv!(text: params[:cv_text], file: params[:file])
    render json: { job_board_entry: serialize(@job_board_entry, with: JobBoardEntryBlueprint) }
  end

  def upload_cover_letter
    @job_board_entry.upload_cover_letter!(text: params[:cover_letter_text], file: params[:file])
    render json: { job_board_entry: serialize(@job_board_entry, with: JobBoardEntryBlueprint) }
  end

  private

  def set_job_board_entry
    @job_board_entry = JobBoardEntry.find_owned!(
      current_user,
      params[:id],
      unauthorized_message: UNAUTHORIZED_MESSAGES.fetch(action_name.to_sym)
    )
  end

  def entry_attributes
    permitted = params.permit(
      :title, :company, :location, :salary, :url, :description, :status, :number, :closing_date
    ).to_h.symbolize_keys
    permitted.delete(:closing_date) unless params.key?(:closing_date)
    permitted
  end

  def import_attributes(entries)
    entries.map do |entry|
      entry.permit(:title, :company, :location, :salary, :url, :description, :status).to_h.symbolize_keys
    end
  end
end
